File: solution/169MajorityElement/169MajorityElement.cpp
#include <algorithm>
#include <iostream>
#include <limits>
#include <sstream>
#include <stack>
#include <unordered_map>
#include <vector>

using namespace std;

/*
Given an array of size n, find the majority element. The majority element is the
element that appears more than ⌊ n/2 ⌋ times.

You may assume that the array is non-empty and the majority element always exist
in the array.

Credits:
Special thanks to @ts for adding this problem and creating all test cases.
*/

class Solution {
public:
  int majorityElement(vector<int> &num) {
    unordered_map<int, int> memap;
    int halfn = num.size() / 2;
    for (const int &x : num) {
      if (memap.find(x) != memap.end()) {
        memap[x]++;
        if (memap[x] > halfn)
          return x;
      } else {
        memap[x] = 1;
      }
    }

    return num[0];
  }
  int majorityElement(vector<int> &num) {
    int n = nums.size();
    int m;
    int i = 0;
    for (auto x : nums) {
      if (i == 0) {
        m = x;
        i = 1;
      } else if (m == x) {
        i++;
      } else {
        i--;
      }
    }
    return m;
  }
};

int main() {
  Solution s;

  return 0;
}

File: src/169MajorityElement/169MajorityElement.rs
impl Solution {
    pub fn majority_element(nums: Vec<i32>) -> i32 {
        let mut nums = nums;
        nums.sort();
        nums[nums.len()/2]
    }
}

impl BitSolution {
    pub fn majority_element(nums: Vec<i32>) -> i32 {
        let mut counts = vec![0;32];
        for num in &nums{
            for i in 0..32{
                counts[i] += (num>>i) & 1;
            }    
        }
        let mut result = 0;
        let half_n = nums.len() as i32/2;
        for i in 0..32{
            let bit = if counts[i] > half_n {1} else {0};
            result |= bit << i; 
        }
        result
    }
}
